Detect attempt to use var-args in closure

estebank opened the pull request:
```
error: unexpected `...`
  --> $DIR/no-closure.rs:11:14
   |
LL |     let f = |...| {};
   |              ^^^ not a valid pattern
   |
   = note: only `extern "C"` and `extern "C-unwind"` functions may have a C variable argument list

error: unexpected `...`
  --> $DIR/no-closure.rs:16:17
   |
LL |     let f = |_: ...| {};
   |                 ^^^
   |
   = note: only `extern "C"` and `extern "C-unwind"` functions may have a C variable argument list
```

Fix rust-lang/rust#146489, when trying to use c-style var-args in a closure. We emit a more targeted message. We also silence inference errors when the pattern is `PatKind::Err`.

estebank committed:
Detect attempt to use var-args in closure

```
error: unexpected `...`
  --> $DIR/varargs-in-closure-isnt-supported.rs:5:20
   |
LL |     let mut lol = |...| ();
   |                    ^^^ not a valid pattern
   |
   = note: C-variadic type `...` is not allowed here
```

estebank committed:
Silence inference error on `PatKind::Err`

estebank committed:
Fix existing test

lcnr commented on tests/ui/c-variadic/no-closure.stderr:
```diff
@@ -10,26 +10,14 @@ error: unexpected `...`
-   |
+   = note: C-variadic type `...` is not allowed here
 
 error[E0743]: C-variadic type `...` may not be nested inside another type
```
existing, but what does "may not be nested inside another type" mean here :sweat_smile:

lcnr commented:
r=me though still still doesn't seem as nice as it could be :thinking: 

could we explicitly state "closures must not be `c-variadic`?

estebank committed:
fmt

estebank committed:
remove redundant test

estebank committed:
Reword note

estebank committed:
Detect top-level `...` in argument type

When writing something like the expression `|_: ...| {}`, we now detect the `...` during parsing explicitly instead of relying on the detection in `parse_ty_common` so that we don't talk about "nested `...` are not supported".

```
error: unexpected `...`
  --> $DIR/no-closure.rs:6:35
   |
LL | const F: extern "C" fn(...) = |_: ...| {};
   |                                   ^^^
   |
   = note: only `extern "C"` and `extern "C-unwind"` functions may have a C variable argument list
```

estebank commented:
> still doesn't seem as nice as it could be 🤔
>
> could we explicitly state "closures must not be c-variadic?

Being that specific will require more special casing in the parser. The new output is at least an improvement :)

```
error: unexpected `...`
  --> $DIR/no-closure.rs:11:14
   |
LL |     let f = |...| {};
   |              ^^^ not a valid pattern
   |
   = note: only `extern "C"` and `extern "C-unwind"` functions may have a C variable argument list

error: unexpected `...`
  --> $DIR/no-closure.rs:16:17
   |
LL |     let f = |_: ...| {};
   |                 ^^^
   |
   = note: only `extern "C"` and `extern "C-unwind"` functions may have a C variable argument list
```

estebank commented on compiler/rustc_trait_selection/src/error_reporting/infer/need_type_info.rs:
```diff
@@ -652,6 +652,10 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
         };
         *err.long_ty_path() = long_ty_path;
+        if let InferSourceKind::ClosureArg { kind: PatKind::Err(_), .. } = kind {
+            // We will have already emitted an error about this pattern.
+            err.downgrade_to_delayed_bug();
```
I guess the rationale for `downgrade_to_delayed_bug` itself not taking an `ErrorGuarantee` is because itself is kind of an error guarantee already :)

The guarantee is more critical for anything that could cause a binary to be emitted otherwise.

lcnr commented:
The PR looks good. For future PRs, I do feel like the change was substantial enough (especially by introducing a check at a new place) that I would have liked to look at it again myself
